File: backend/common/export.py
import csv
from django.http import HttpResponse
from datetime import datetime


def export_as_csv(queryset, fields, field_labels=None, filename=None):
    if filename is None:
        filename = f'export_{datetime.now().strftime("%Y%m%d_%H%M%S")}.csv'
    response = HttpResponse(content_type='text/csv; charset=utf-8-sig')
    response['Content-Disposition'] = f'attachment; filename="{filename}"'
    writer = csv.writer(response)
    headers = [field_labels.get(f, f) for f in fields] if field_labels else list(fields)
    writer.writerow(headers)
    for obj in queryset:
        row = []
        for field in fields:
            val = getattr(obj, field, '')
            if callable(val):
                val = val()
            if hasattr(val, 'all'):
                val = ', '.join(str(v) for v in val.all())
            row.append(str(val) if val is not None else '')
        writer.writerow(row)
    return response


def validate_and_export(queryset, fields, required_fields=None, **kwargs):
    if required_fields:
        for obj in queryset:
            for f in required_fields:
                val = getattr(obj, f, None)
                if val is None or val == '':
                    raise ValueError(f'对象 {obj.pk} 缺少必填字段: {f}')
    return export_as_csv(queryset, fields, **kwargs)

File: src/routes/api/audit/export.ts
import { getDB, initDB } from "~/server/db";
import type { APIEvent } from "@solidjs/start/server";

initDB();

export async function GET(event: APIEvent) {
  const db = getDB();
  const url = new URL(event.request.url);
  const startDate = url.searchParams.get("startDate");
  const endDate = url.searchParams.get("endDate");
  const format = url.searchParams.get("format") || "json";
  const now = Date.now();
  const start = startDate ? parseInt(startDate) : now - 30 * 24 * 60 * 60 * 1000;
  const end = endDate ? parseInt(endDate) : now;

  const logs = db
    .prepare(
      "SELECT * FROM audit_logs WHERE created_at BETWEEN ? AND ? ORDER BY created_at DESC"
    )
    .all(start, end);

  if (format === "csv") {
    const headers = ["ID", "操作", "实体类型", "实体ID", "操作人", "角色", "备注", "时间"];
    const rows = logs.map((log: any) => [
      log.id,
      log.action,
      log.entity_type,
      log.entity_id,
      log.operator,
      log.operator_role,
      log.remark || "",
      new Date(log.created_at).toLocaleString("zh-CN"),
    ]);

    const csv = [headers, ...rows].map(row => row.map(cell => `"${cell}"`).join(",")).join("\n");
    return new Response("\ufeff" + csv, {
      headers: {
        "Content-Type": "text/csv; charset=utf-8",
        "Content-Disposition": `attachment; filename="audit-log-${Date.now()}.csv"`,
      },
    });
  }

  return Response.json(logs);
}
